_source/_posts/2022-01-26-webassembly-on-kubernetes-with-rust.md

```diff
@@ -0,0 +1,365 @@
+
+```bash
+rustup target add wasm32-wasi
+```
+
+Now you can create a new Rust application with cargo.
+
+```bash
+cargo new --bin rust-wasm
+```
+
+Open the created `rust-wasm` folder in your favorite IDE. I use Visual Studio Code with the fantastic [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=matklad.rust-analyzer) and [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) plugins to develop in Rust.
+
+## Create the WebAssembly workload
+
+Let's write a small service that will print random cat facts on the console. For this, you could use a free public [API that provides random cat facts](https://catfact.ninja/fact) for this.
+
+Edit `cargo.toml` and add the following dependencies:
+
+```toml
+[dependencies]
+wasi-experimental-http = "0.7"
+http = "0.2.5"
+serde_json = "1.0.74"
+env_logger = "0.9"
+log = "0.4"
+```
+
+Then edit `src/main.rs` and add the following code:
+
+```rust
+use http;
+use serde_json::Value;
+use std::{str, thread, time};
+
+fn main() {
+    env_logger::init();
+    let url = "https://catfact.ninja/fact".to_string();
+    loop {
+        let req = http::request::Builder::new()
+            .method(http::Method::GET)
+            .uri(&url)
+            .header("Content-Type", "text/plain");
+        let req = req.body(None).unwrap();
+
+        log::debug!("Request: {:?}", req);
+
+        // send request using the experimental bindings for http on wasi
+        let mut res = wasi_experimental_http::request(req).expect("cannot make request");
+
+        let response_body = res.body_read_all().unwrap();
+        let response_text = str::from_utf8(&response_body).unwrap().to_string();
+        let headers = res.headers_get_all().unwrap();
+
+        log::debug!("{}", res.status_code);
+        log::debug!("Response: {:?} {:?}", headers, response_text);
+
+        // parse the response to json
+        let cat_fact: Value = serde_json::from_str(&response_text).unwrap();
+
+        log::info!("Cat Fact: {}", cat_fact["fact"].as_str().unwrap());
+
+        thread::sleep(time::Duration::new(60, 0));
+    }
+}
+```
+
+The code is simple. It makes a GET request to the API and parses and prints the response every 60 seconds. Now you can build this into a Wasm binary using the below cargo command:
+
+```bash
+cargo build --release --target wasm32-wasi
+```
+
+That's it. You have successfully created a WebAssembly binary using Rust.
+
+## Run the workload locally (optional)
+
+Let's run the workload locally using [wasmtime](https://wasmtime.dev), a small JIT-style runtime for Wasm and WASI. Since wasmtime doesn't support networking out of the box, we need to use the [wrapper](https://github.com/deislabs/wasi-experimental-http#testing-using-the-wasmtime-http-binary) provided by [wasi-experimental-http](https://github.com/deislabs/wasi-experimental-http). You can build it from source it using the below command.
```

Comment:
```suggestion
Let's run the workload locally using [Wasmtime](https://wasmtime.dev), a small JIT-style runtime for Wasm and WASI. Since Wasmtime doesn't support networking out of the box, we need to use the [wrapper](https://github.com/deislabs/wasi-experimental-http#testing-using-the-wasmtime-http-binary) provided by [wasi-experimental-http](https://github.com/deislabs/wasi-experimental-http). You can build it from source using the below command.
```
Capitalize Wasmtime. remove extra "it"